File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.js
let cache = new Map();

function LRU(city,i){
    let arr = [...cache];
    arr.sort((a,b)=>a[1] - b[1]);
    cache.delete(arr[0][0]);
    cache.set(city,i);
}

function solution(cacheSize, cities) {
    var answer = 0;
    let remain = cacheSize;
    
    for(let i = 0; i<cities.length; i++){
        cities[i] = cities[i].toLowerCase();
        if(cache.has(cities[i])) {
            cache.set(cities[i],i);
            answer +=1;
        }
        else{
            if(remain > 0){
                cache.set(cities[i],i);
                remain--;
            }
            else{
                if(cacheSize > 0) LRU(cities[i],i);
            }
            answer += 5;
        }
        
    }
    return answer;
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        Map<String, Integer> cache = new HashMap<>();
        Deque<String> queue = new ArrayDeque<>();
        
        int answer = 0;
        
        for(int i = 0; i < cities.length; i++){
            String city = cities[i].toLowerCase();
            
            if(cacheSize == 0){ //캐시 공간 없음
                answer += 5;
                continue;
            } else if(cache.containsKey(city)){ //캐시 히트
                answer += 1;
            } else if(cache.size() < cacheSize){  //캐시 공간 있음
                answer += 5;
            } else if(cache.size() == cacheSize){
                answer += 5;
                int min = i;
                String delCity = "";
                for(var cityInCache: cache.entrySet()){
                    if(min > cityInCache.getValue()) {
                        delCity = cityInCache.getKey();
                        min = cityInCache.getValue();
                    }
                }
                cache.remove(delCity);
            }
            cache.put(city, i);
        }
        
        return answer;
    }
}